Add container environment variables table

alongoldboim opened the pull request:
ref : https://github.com/ManageIQ/manageiq/pull/3327

alongoldboim commented:
@simon3z Added a table with the env variables and an expanding functionality. (can be used application wise by using the same table partial)
![f773c_](https://cloud.githubusercontent.com/assets/11405684/12074869/d35196aa-b174-11e5-87ae-1aa319df82aa.gif)

DNNX commented on app/assets/javascripts/miq_application.js:
```diff
@@ -1383,3 +1383,9 @@ function miqInitMainContent() {
+
+function expanded_td(link) {
+    var link = $(link);
+    link.find("i").toggleClass("glyphicon-plus").toggleClass("glyphicon-minus");
```
These 2 toggleClass calls can be replaced with one: `link.find("i").toggleClass("glyphicon-plus glyphicon-minus"))`.

[JQuery docs](http://api.jquery.com/toggleclass/):

> This method takes **one or more class names** as its parameter. In the first version, if an element in the matched set of elements already has the class, then it is removed; if an element does not have the class, then it is added.

DNNX commented on app/views/shared/summary/_textual_multilabel.html.haml:
```diff
@@ -1,5 +1,5 @@
 - if items && items[:values].present?
-  %table.table.table-bordered.table-striped
+  %table.table.table-bordered.table-striped{:class => items[:additional_table_class] ? items[:additional_table_class] : ""}
```
`{:class => items[:additional_table_class] || ""}`

DNNX commented on app/views/shared/summary/_textual_multilabel.html.haml:
```diff
@@ -12,4 +12,9 @@
         - item.each do |value|
           %td
-            = value
+            %div{:class => "pull-left"}
```
Can we use `.pull-left` here instead of `%div{:class => "pull-left"}`?

DNNX commented on app/views/shared/summary/_textual_multilabel.html.haml:
```diff
@@ -12,4 +12,9 @@
+            %div{:class => "pull-left"}
+              = value
+            - if items[:expandable] && value.length > 40
+              %div{:class => "pull-right"}
```
And here too.

DNNX commented on app/views/shared/summary/_textual_multilabel.html.haml:
```diff
@@ -12,4 +12,9 @@
+            - if items[:expandable] && value.length > 40
+              %div{:class => "pull-right"}
+                %a{:onclick => "expanded_td(this)"}
+                  %i{:class => 'glyphicon glyphicon-plus'}
```
`%i.glyphicon.glyphicon-plus` maybe?

alongoldboim commented:
@DNNX Thanks for the comments, still need to get used to haml :+1:

abonas commented:
@alongoldboim, couple of comments on look and feel of the table based on the screenshot:
1. Type column at the moment seems redundant. All env variables have "value" in the type column.
2. The expand/collapse done with +/- is misleading. Typically +/- are used for add/remove.
3. If the Type column is here to stay, it is too wide. I suggest making it narrower, making more room for the value column.
4. The value column name needs to be capitalized like other column names.
5. Table title seems partial  - "Environment variables" will be more clear than just "Environment".

abonas commented on app/helpers/container_helper/textual_summary.rb:
```diff
@@ -109,4 +109,23 @@ def textual_se_linux_level
+    @record.container_definition.container_env_vars.collect do |var|
+      [
+        var.name,
+        var.value.nil? ? "REFERENCE" : "VALUE",
```
what is the meaning of Reference? And why is this part hardcoded?

moolitayer replied:
The value of an env var can be a string (that includes other vars) called value OR a reference to other fields called valueFrom[1]

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L701

abonas replied:
> The value of an env var can be a string (that includes other vars) called value OR a reference to other fields called valueFrom[1]
> 
> [1] https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L701

I wonder if it still requires hardcoded strings on miq side or it can be somehow generically extracted from k8s output

moolitayer replied:
since we are using names different from k8s' valueFrom and fieldRef I can't think of a way

alongoldboim commented:
@abonas regarding you comments:
1) Its redudant only in this exmaple. 
2) I can change to fa-angle-right and fa-angle-down, do you find it better?
3) Fixed, would actually like to replace text with icons but didn't find appropriate ones.
4) Fixed.
5) Fixed.
6) The table will increase its height.

moolitayer commented on app/views/container/_container_show.html.haml:
```diff
@@ -7,3 +7,7 @@
+.row
+  .col-sm-12.col-md-12.col-lg-6
+    = render :partial => "shared/summary/textual_multilabel", :locals => {:title => _("Environment"),
+                                                                          :items => textual_group_env}
```
new line missing

abonas commented:
> @abonas Thanks for the comments:

Please use 'reply to' otherwise there's a lot of scrolling up/down to understand which comment/item you are referring to.

> 1) Its redudant only in this exmaple. 

can you elaborate more what do you mean?
> 2) I can change to fa-angle-right and fa-angle-down, do you find it better?

yes
> 3) Fixed, would actually like to replace text with icons but didn't find appropriate ones.\

it's currently not the standard in miq summary pages, but I agree that it could be nice.
> 4) Fixed.
> 5) Fixed.
> 6) The table will increase its height.

which means the entire page will be scrollable, is that correct?

moolitayer commented:
@alongoldboim please attach an image with (expanded) long environment variables.
If you are using molecule take a look at a container named 'registry'.

abonas commented:
please also note, we are supposed to work with Patternfly look and feel / components.
So for example, the relevant icons should be from here: https://www.patternfly.org/styles/icons/

moolitayer commented on app/assets/javascripts/miq_application.js:
```diff
@@ -1383,3 +1383,9 @@ function miqInitMainContent() {
   $('#main-content').css('height', 'calc(100% - ' + height + 'px)')
 }
+
+function expanded_td(link) {
```
toggle_expention ?

moolitayer commented:
@miq-bot add_label ui, providers/containers, enhancement

alongoldboim commented:
Regarding your comments:
1) In the example i have taken there were no environment variables with type = REFERENCE, in general there could be.
2) Fixed.
3) The entire page will be scrollable.
4) Screenshot was update.

moolitayer commented on app/assets/javascripts/miq_application.js:
```diff
@@ -1383,3 +1383,9 @@ function miqInitMainContent() {
+function expanded_td(link) {
+    var link = $(link);
+    link.find("i").toggleClass("glyphicon-plus glyphicon-minus");
+    link.parent().parent().children(0).toggleClass("expanded");
```
I prefer: parent().parent() => parent('td')

moolitayer commented on app/assets/javascripts/miq_application.js:
```diff
@@ -1383,3 +1383,9 @@ function miqInitMainContent() {
+function toggle_expansion(link) {
+    var link = $(link);
+    link.find("i").toggleClass("fa-angle-right fa-angle-down");
+    link.parent().parent().children(0).toggleClass("expanded");
```
I prefer: parent().parent() => parent('td')

alongoldboim replied:
parent('td') won't work here, can use parentsUntil() or closest() (went with closest).

moolitayer commented:
@alongoldboim note this currently does not fix #5617. Css selectors in this patch are applied only to the table created for environment using 'table-fixed' class so the other examples from #5617 using textual multilabel are not fixed. 

I think one way to change this is: in case the text is not expandable (items[:expandable] => :false) add the text in a class that will enable these changes ( including div.expanded to have the text always expanded).

Other then that this issue pr looks good to me now. Adding @epwinchell that addressed similar issues before.

alongoldboim commented:
@moolitayer didn't mean to fix #5617(refed it by mistake, meant to ref this https://github.com/ManageIQ/manageiq/pull/3327).
if we do want to fix #5617 i advise doing it in a separate pr dedicated to that problem.

moolitayer commented:
> @moolitayer didn't mean to fix #5617(refed it by mistake, meant to ref this #3327).
> If we do want to fix #5617 i advise doing it in a separate pr dedicated to that problem.

In that case a separate pr is probably a good idea.

simon3z commented on app/views/shared/summary/_textual_multilabel.html.haml:
```diff
@@ -12,4 +12,11 @@
         - item.each do |value|
           %td
-            = value
+            - if value && items[:expandable] && value.length > 50
```
@alongoldboim I hate hard-coded values... any idea on how to drop this empirical value (50) in favor of something more automated? Is it something that the js side can decide depending on the cell size? Maybe you can find something interesting here: http://www.google.com/search?q=jquery%20cell%20overflow

alongoldboim commented:
@simon3z
* Removed hard coded values.
* Added a js function to evaluate the needed table cells.

simon3z commented:
> @simon3z
> 
> Removed hard coded values.
> Added a js function to evaluate the needed table cells.

Thanks! LGTM :+1:

@alongoldboim do you need to update the screenshot? Make it easy to understand what's the final look (arrows or +/- signs, etc.) and then get @serenamarie125 approval.

@zeari @abonas please review.

alongoldboim commented:
@simon3z screenshot was updated.

simon3z commented:
> @simon3z screenshot was updated.

@alongoldboim please contact @serenamarie125 to get her review.

epwinchell commented:
LGTM

simon3z commented:
@alongoldboim please rebase, cc @chessbyte ready for merge

simon3z commented:
Ping @chessbyte LGTM :+1: ready for merge

chessbyte commented:
@dclarizio @martinpovolny @h-kataria Would you please review as this PR is mostly in the UI code.

alongoldboim commented:
@dclarizio after this will be merged, i can make this an application wide change for all tables, is that something we would be interested in?

dclarizio commented:
@alongoldboim please rebase, I think this is good to go.

alongoldboim commented:
@dclarizio done.
